dashboard/app/assets/javascripts/header.js.erb

```diff
@@ -2,110 +2,575 @@
-    }
-    progressContainer.append(div).append('\n');
-  });
-  $('.level_free_play').qtip({
-    content: {
-      attr: 'title'
-    },
-    position: {
-      my: 'top center',
-      at: 'bottom center'
-    }
-  });
-
-  function showHeaderPopup(target) {
-    $('.header_popup').show();
-    $('.header_popup_link_glyph').html('&#x25B2;');
-    $('.header_popup_link_text').text('<%= I18n.t(:less) %>');
-    $(document).on('click', hideHeaderPopup);
-    lazyLoadPopup($(target).closest('.header_trophy_link').length > 0);
-  }
-  function hideHeaderPopup() {
-    $('.header_popup').hide();
-    $('.header_popup_link_glyph').html('&#x25BC;');
-    $('.header_popup_link_text').text('<%= I18n.t(:more) %>');
-    $(document).off('click', hideHeaderPopup);
-  }
-  $('.header_popup_link, .header_trophy_link').click(function (e) {
-    e.stopPropagation();
-    $('.header_popup').is(':visible') ? hideHeaderPopup() : showHeaderPopup(e.target);
-  });
-  $('.header_popup').click(function (e) {
-    e.stopPropagation(); // Clicks inside the popup shouldn't close it
-  });
-
-  var popupLoaded = false;
-  function lazyLoadPopup(trophiesClicked) {
-    if (!popupLoaded) {
-      popupLoaded = true;
-      $.ajax({
-        url: scriptData.statsPath,
-        data: {
-          script_id: scriptData.scriptId,
-          script_level_id: scriptData.scriptLevelId
-        }, success: function (result) {
-          $('.header_popup_body').html(result);
-          if (trophiesClicked) {
-            jumpToTrophies();
-          }
+
+// ****************
+//  IMPORTANT NOTE
+// ****************
+//
+// This file contains proof-of-concept code (though it's safe for production)
+// - UIComponent is an example React implementation
+// - UIStore is an example Flux implementation
+// - UIRouter is a really simple router
+//
+// These classes (and the derived ones) look pretty 'bulky' as they are written today.
+// If we begin using React, this file will shrink to less than half of its current size.
+// With CoffeeScript it'll shrink at least another half.
+// There's really not that much "logic" here, it's just boilerplate that can be eliminated!
+
+
+// New private scope -- a lot of this is boilerplate for class structures, taken from coffeescript
+(function() {
+  var
+    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
+    __hasProp = {}.hasOwnProperty,
+    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
+    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };
+
+  // class UIComponent: this is a lightweight inheritable class modeled after React
+  UIComponent = (function() {
```

Comment:
Wow that's wild, I wonder why.  What's the point of having a closure if the class definitions get leaked anyway?  :)